### PR TITLE
 Enabling T5ForConditionalGeneration Inference using static shapes

### PR DESCRIPTION
**Pull request by bhargaveede**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by ZhaiFeiyue** [comment]
@bhargaveede very nice PR, since there are lots of recompilation for `run_translation` with T5, so is this PR support translation?
the cmd for translation is
```bash
python run_translation.py \
    --model_name_or_path t5-small \
    --do_eval \
    --source_lang en \
    --target_lang ro \
    --source_prefix "translate English to Romanian: " \
    --dataset_name wmt16 \
    --dataset_config_name ro-en \
    --output_dir /tmp/tst-translation \
    --per_device_eval_batch_size 4 \
    --overwrite_output_dir \
    --predict_with_generate \
    --use_habana \
    --use_lazy_mode \
    --use_hpu_graphs_for_inference \
    --gaudi_config_name Habana/t5 \
    --ignore_pad_token_for_loss False \
    --pad_to_max_length
```

File: optimum/habana/transformers/generation/utils.py [hunk 1]
```diff
@@ -1260,7 +1280,8 @@ def greedy_search(
                 else:
                     # case3 (default case): token_idx is None
                     next_tokens_scores = logits_processor(input_ids, next_token_logits)
-
+            # generate_end = time.perf_counter()
+            # print("Time", generate_end - generate_start)
```
[the remark this answers is not in the document]
Comment (bhargaveede) [reply optimum/habana/transformers/generation/utils.py:1284]: This is in Draft. We're working on few internal issues. Will upload patch with clean code

File: examples/summarization/run_summarization.py [hunk 2]
```diff
@@ -475,13 +476,24 @@ def main():
         trust_remote_code=model_args.trust_remote_code,
         use_cache=False if training_args.gradient_checkpointing else model_args.use_cache,
     )
+    is_bart = True if model_args.model_name_or_path in [
+            "facebook/bart-large-mnli",
+            "facebook/bart-large-cnn",
+            "facebook/bart-large",
+            "facebook/bart-base",
+            "facebook/bart-large-xsum",
+        ] else False
```
[the remark this answers is not in the document]
Comment (bhargaveede) [reply examples/summarization/run_summarization.py:485]: @ssarkar2 Right, I'll do the change

**Comment by bhargaveede** [comment]
> @bhargaveede very nice PR, since there are lots of recompilation for `run_translation` with T5, so is this PR support translation? the cmd for translation is
> 
> ```shell
> python run_translation.py \
>     --model_name_or_path t5-small \
>     --do_eval \
>     --source_lang en \
>     --target_lang ro \
>     --source_prefix "translate English to Romanian: " \
>     --dataset_name wmt16 \
>     --dataset_config_name ro-en \
>     --output_dir /tmp/tst-translation \
>     --per_device_eval_batch_size 4 \
>     --overwrite_output_dir \
>     --predict_with_generate \
>     --use_habana \
>     --use_lazy_mode \
>     --use_hpu_graphs_for_inference \
>     --gaudi_config_name Habana/t5 \
>     --ignore_pad_token_for_loss False \
>     --pad_to_max_length
> ```

@ZhaiFeiyue  Modeling changes related to t5 will be common for both translation and summarization.
In generate() call of utils.py We set static_shapes based on model and that'll help in reducing model decoder recompilations.

So, even for translation, you should see reduced recompilations and improved perf

**Comment by bhargaveede** [comment]
@regisss I've rebased it to latest.
I verified many tests on my side and working on few fixes.
Meanwhile, can you start reviewing the code so that if we need any major changes, we can change them now.

**Review by regisss — COMMENTED** [review]
Here is a partial review, I'll do the rest of the `beam_search` method and T5 modeling on Monday

File: optimum/habana/transformers/models/t5/modeling_t5.py [hunk 3]
```diff
@@ -24,14 +59,797 @@ def gaudi_t5_layernorm_forward(self, hidden_states):
...
+    >>> logits = outputs.logits
+
+    >>> # inference
+    >>> input_ids = tokenizer(
+    ...     "summarize: studies have shown that owning a dog is good for you", return_tensors="pt"
+    ... ).input_ids  # Batch size 1
+    >>> outputs = model.generate(input_ids)
+    >>> print(tokenizer.decode(outputs[0], skip_special_tokens=True))
+    >>> # studies have shown that owning a dog is good for you.
+    ```"""
```
Comment (regisss) [inline optimum/habana/transformers/models/t5/modeling_t5.py:687]: Let's remove this docstring as it is exactly the same as in Transformers

File: optimum/habana/transformers/models/t5/modeling_t5.py [hunk 4]
```diff
@@ -24,14 +59,797 @@ def gaudi_t5_layernorm_forward(self, hidden_states):
...
+    use_cache = use_cache if use_cache is not None else self.config.use_cache
+    return_dict = return_dict if return_dict is not None else self.config.use_return_dict
```
Comment (regisss) [inline optimum/habana/transformers/models/t5/modeling_t5.py:689]: Add a blank line below

File: optimum/habana/transformers/models/t5/modeling_t5.py [hunk 5]
```diff
@@ -24,14 +59,797 @@ def gaudi_t5_layernorm_forward(self, hidden_states):
...
+        encoder_outputs = BaseModelOutput(
+            last_hidden_state=encoder_outputs[0],
+            hidden_states=encoder_outputs[1] if len(encoder_outputs) > 1 else None,
+            attentions=encoder_outputs[2] if len(encoder_outputs) > 2 else None,
+        )
+
+    hidden_states = encoder_outputs[0]
+
+    if self.model_parallel:
+        torch.cuda.set_device(self.decoder.first_device)
```
Comment (regisss) [inline optimum/habana/transformers/models/t5/modeling_t5.py:718]: To remove

File: optimum/habana/transformers/models/t5/modeling_t5.py [hunk 6]
```diff
@@ -24,14 +59,797 @@ def gaudi_t5_layernorm_forward(self, hidden_states):
...
+    # Set device for model parallelism
+    if self.model_parallel:
+        torch.cuda.set_device(self.decoder.first_device)
+        hidden_states = hidden_states.to(self.decoder.first_device)
+        if decoder_input_ids is not None:
+            decoder_input_ids = decoder_input_ids.to(self.decoder.first_device)
+        if attention_mask is not None:
+            attention_mask = attention_mask.to(self.decoder.first_device)
+        if decoder_attention_mask is not None:
+            decoder_attention_mask = decoder_attention_mask.to(self.decoder.first_device)
```
Comment (regisss) [inline optimum/habana/transformers/models/t5/modeling_t5.py:733]: To remove

File: optimum/habana/transformers/models/t5/modeling_t5.py [hunk 7]
```diff
@@ -24,14 +59,797 @@ def gaudi_t5_layernorm_forward(self, hidden_states):
...
+        max_output_length=max_output_length,
+    )
+
+    sequence_output = decoder_outputs[0]
+
+    # Set device for model parallelism
+    if self.model_parallel:
+        torch.cuda.set_device(self.encoder.first_device)
+        self.lm_head = self.lm_head.to(self.encoder.first_device)
+        sequence_output = sequence_output.to(self.lm_head.weight.device)
```
Comment (regisss) [inline optimum/habana/transformers/models/t5/modeling_t5.py:759]: To remove

File: optimum/habana/transformers/models/t5/modeling_t5.py [hunk 8]
```diff
@@ -24,14 +59,797 @@ def gaudi_t5_layernorm_forward(self, hidden_states):
...
+    input_ids,
+    past_key_values=None,
+    attention_mask=None,
+    head_mask=None,
+    decoder_head_mask=None,
+    cross_attn_head_mask=None,
+    use_cache=None,
+    encoder_outputs=None,
+    token_idx=None,
+    decoder_attention_mask=None,
```
Comment (regisss) [inline optimum/habana/transformers/models/t5/modeling_t5.py:804]: Can we keep the same order as in Transformers? https://github.com/huggingface/transformers/blob/b71f20a7c9f3716d30f6738501559acf863e2c5c/src/transformers/models/t5/modeling_t5.py#L1800

File: optimum/habana/transformers/models/t5/modeling_t5.py [hunk 9]
```diff
@@ -24,14 +59,797 @@ def gaudi_t5_layernorm_forward(self, hidden_states):
...
+        "max_output_length": max_output_length,
+    }
+
+
+def gaudi_T5ForConditionalGeneration_reorder_cache(self, past_key_values, beam_idx):
+    # if decoder past is not included in output
+    # speedy decoding is disabled and no need to reorder
+    if past_key_values is None:
+        logger.warning("You might want to consider setting `use_cache=True` to speed up decoding")
+        return past_key_values
```
Comment (regisss) [inline optimum/habana/transformers/models/t5/modeling_t5.py:835]: Add a blank line below

File: optimum/habana/transformers/models/t5/modeling_t5.py [hunk 10]
```diff
@@ -24,14 +59,797 @@ def gaudi_t5_layernorm_forward(self, hidden_states):
...
+        # batch dim of `past` is at 2nd position
+        reordered_layer_past_states = ()
+        for layer_past_state in layer_past_states:
+            # need to set correct `past` for each of the four key / value states
+            reordered_layer_past_states = reordered_layer_past_states + (
+                layer_past_state.index_select(0, beam_idx.to(layer_past_state.device)),
+            )
+
+        assert reordered_layer_past_states[0].shape == layer_past_states[0].shape
+        assert len(reordered_layer_past_states) == len(layer_past_states)
```
Comment (regisss) [inline optimum/habana/transformers/models/t5/modeling_t5.py:848]: Replace these two asserts: https://github.com/huggingface/transformers/blob/b71f20a7c9f3716d30f6738501559acf863e2c5c/src/transformers/models/t5/modeling_t5.py#L1850

File: optimum/habana/transformers/generation/utils.py [hunk 11]
```diff
@@ -2016,7 +2110,7 @@ def beam_search(
                 next_token_logits, dim=-1
             )  # (batch_size * num_beams, vocab_size)
 
-            next_token_scores_processed = logits_processor(input_ids, next_token_scores)
+            next_token_scores_processed = logits_processor(input_ids[:, :token_idx], next_token_scores)
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:2113]: What if `token_idx` is None?

File: optimum/habana/transformers/generation/utils.py [hunk 12]
```diff
@@ -479,7 +482,7 @@ def generate(
             model_kwargs["use_cache"] = True
         else:
             model_kwargs["use_cache"] = generation_config.use_cache
-
+        model_kwargs["max_output_length"] = generation_config.max_length or generation_config.max_new_tokens
```
[the remark this answers is not in the document]
Comment (bhargaveede) [reply optimum/habana/transformers/generation/utils.py:495]: We use this in the modeling_t5.py

File: optimum/habana/transformers/models/t5/modeling_t5.py [hunk 13]
```diff
@@ -24,14 +59,797 @@ def gaudi_t5_layernorm_forward(self, hidden_states):
...
+    return outputs  # hidden-states, present_key_value_states, (self-attention position bias), (self-attention weights), (cross-attention position bias), (cross-attention weights)
+
+
+# Warning message for FutureWarning: head_mask was separated into two input args - head_mask, decoder_head_mask
+__HEAD_MASK_WARNING_MSG = """
+The input argument `head_mask` was split into two arguments `head_mask` and `decoder_head_mask`. Currently,
+`decoder_head_mask` is set to copy `head_mask`, but this feature is deprecated and will be removed in future versions.
+If you do not want to use any `decoder_head_mask` now, please set `decoder_head_mask = torch.ones(num_layers,
+num_heads)`.
+"""
```
[the remark this answers is not in the document]
Comment (bhargaveede) [reply optimum/habana/transformers/models/t5/modeling_t5.py:633]: We use this in gaudi_T5ForConditionalGeneration_forward

File: examples/summarization/run_summarization.py [hunk 14]
```diff
@@ -491,6 +493,7 @@ def main():
         revision=model_args.model_revision,
         token=model_args.token,
         trust_remote_code=model_args.trust_remote_code,
+        torch_dtype=torch.bfloat16 if mixed_precision else torch.float32,
```
[the remark this answers is not in the document]
Comment (bhargaveede) [reply examples/summarization/run_summarization.py:497]: Casting the model gave better results during inference. Shall I add inference check?

File: optimum/habana/transformers/generation/utils.py [hunk 15]
```diff
@@ -437,6 +439,7 @@ def generate(
         generation_config = copy.deepcopy(generation_config)
         if generation_config.static_shapes is None:
             generation_config.static_shapes = self.config.model_type in MODELS_OPTIMIZED_WITH_STATIC_SHAPES
+            self.generation_config.static_shapes = generation_config.static_shapes
```
[the remark this answers is not in the document]
Comment (bhargaveede) [reply optimum/habana/transformers/generation/utils.py:451]: Yes

**Comment by bhargaveede** [comment]
@regisss Can you review the changes?

**Comment by regisss** [comment]
> @regisss Can you review the changes?

Sorry for the delay, I'll review it in the beginning of next week as I'm doing a sprint to enable CI on Gaudi2.

File: examples/summarization/run_summarization.py [hunk 16]
```diff
@@ -482,6 +483,7 @@ def main():
...
```
[the remark this answers is not in the document]
Comment (regisss) [reply examples/summarization/run_summarization.py:487]: We shouldn't have this for the tokenizer

File: examples/summarization/run_summarization.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (regisss) [reply examples/summarization/run_summarization.py:497]: Tranformers has the [`bf16_full_eval`](https://huggingface.co/docs/transformers/v4.35.0/en/main_classes/trainer#transformers.TrainingArguments.bf16_full_eval) training arg to do exactly this. It is currently not enabled in Optimum Habana (see [here](https://github.com/huggingface/optimum-habana/blob/47802e85225833c45ae175bb6ed6db7a0be7fdfe/optimum/habana/transformers/training_args.py#L317)) but we can do that in another PR.
So let's remove this line and we'll enable full bf16 evaluation next.

**Comment by bhargaveede** [comment]
@regisss Ok. I've resolved conflicts and CI is passing as well.

> > @regisss Can you review the changes?
> 
> Sorry for the delay, I'll review it in the beginning of next week as I'm doing a sprint to enable CI on Gaudi2.

**Comment by vivekgoe** [comment]
@regisss reminder to review these changes when you get a chance. Thanks.

**Comment by bhargaveede** [comment]
@regisss I've taken care of the review comments. Can you review and merge

File: optimum/habana/transformers/models/t5/modeling_t5.py [hunk 17]
```diff
@@ -27,3 +45,601 @@ def gaudi_t5_layernorm_forward(self, hidden_states):
...
+    )
+
+
+# # Warning message for FutureWarning: head_mask was separated into two input args - head_mask, decoder_head_mask
+# __HEAD_MASK_WARNING_MSG = """
+# The input argument `head_mask` was split into two arguments `head_mask` and `decoder_head_mask`. Currently,
+# `decoder_head_mask` is set to copy `head_mask`, but this feature is deprecated and will be removed in future versions.
+# If you do not want to use any `decoder_head_mask` now, please set `decoder_head_mask = torch.ones(num_layers,
+# num_heads)`.
+# """
```
Comment (regisss) [inline optimum/habana/transformers/models/t5/modeling_t5.py:505]: To remove

File: optimum/habana/transformers/modeling_utils.py [hunk 18]
```diff
@@ -261,6 +267,19 @@ def adapt_transformers_to_gaudi():
     transformers.models.mpt.modeling_mpt.MptAttention.forward = gaudi_mpt_attention_forward
     transformers.models.mpt.modeling_mpt.MptBlock.forward = gaudi_mpt_block_forward
 
+    # TODO: revisit this when switching to Transformers v4.33
+    # see https://github.com/huggingface/transformers/pull/25394
+    # Hack for running T5 with DeepSpeed Zero-3
+    # Dropout kernel improvement for Flan-T5
```
Comment (regisss) [inline optimum/habana/transformers/modeling_utils.py:273]: This should not be there anymore, see main: https://github.com/huggingface/optimum-habana/blob/3aa707e386f9981e7b0ff5da7c52c87412b9a1e8/optimum/habana/transformers/modeling_utils.py#L257
I think `gaudi_t5_layernorm_forward` got lost at some point when solving merge conflicts.

**Review by regisss — COMMENTED** [review]
There is some old code in `modeling_utils.py` as I wrote in a comment below. Let's fix that, then I'll run the regression test and after that I'll merge!

File: optimum/habana/transformers/modeling_utils.py (hunk 18 above)
Comment (bhargaveede) [reply optimum/habana/transformers/modeling_utils.py:273]: https://github.com/huggingface/optimum-habana/blob/2b3323efbbfa41caab28f7b69a9ff8db15539a2b/optimum/habana/transformers/modeling_utils.py#L262

gaudi_t5_layernorm_forward is there

**Comment by bhargaveede** [comment]
@regisss I have addressed the review comments

File: optimum/habana/transformers/modeling_utils.py (hunk 18 above)
Comment (regisss) [reply optimum/habana/transformers/modeling_utils.py]: My bad I read too quickly

**Review by regisss — COMMENTED** [review]
Thanks, LGTM!
I'm just going to run the regression tests today before merging.

**Comment by regisss** [comment]
@bhargaveede Running the regression tests, I get
```
***** predict metrics *****
  predict_gen_len            =      127.5
  predict_loss               =        nan
  predict_rouge1             =        0.0
  predict_rouge2             =        0.0
  predict_rougeL             =        0.0
  predict_rougeLsum          =        0.0
```
 for T5 on both Gaudi1 and Gaudi2. Can you reproduce it?

**Comment by bhargaveede** [comment]
@regisss Nan is happening after reverting "Model moving to bf16" and bf16 is the run that's part of regression tests.
Looking into it.

**Comment by bhargaveede** [comment]
@regisss I have done a change where Nan is happening due to autocast. Similar fix was done earlier in gaudi_get_extended_attention_mask as well.
Please review and merge.

Also, Is there any existing issue (or) PR open for bf16_full_eval.
I can take a look as that would further improve the performance.

**Review by regisss — APPROVED** [review]
LGTM!
I'm going to launch the regression tests and let's wait for them before merging this PR. The change to `gaudi_get_extended_attention_mask` will impact other models, I'm not worried about it but let's make sure everything is all right.

Regarding `bf16_full_eval`, it has been in my to-do list for a while but I have not got the time to deal with it yet. If you want to take a look at it, what needs to be done is the following:
- remove `bf16_full_eval` from the list [here](https://github.com/huggingface/optimum-habana/blob/94fe9152a4654b25af384dbf097f18ae57f22def/optimum/habana/transformers/training_args.py#L57) and remove the error [here](https://github.com/huggingface/optimum-habana/blob/94fe9152a4654b25af384dbf097f18ae57f22def/optimum/habana/transformers/training_args.py#L317)
- take a look at the occurrences of `bf16_full_eval` in Transformers' Trainer (in particular at lines 1529, 3208 and 3816 [here](https://github.com/huggingface/transformers/blob/v4.34.1/src/transformers/trainer.py)) and replicate them in GaudiTrainer

That should be all that is needed but there might be a bit more, we should see this quite fast if any `bf16_full_eval` is missing anywhere else.

**Comment by bhargaveede** [comment]
> LGTM! I'm going to launch the regression tests and let's wait for them before merging this PR. The change to `gaudi_get_extended_attention_mask` will impact other models, I'm not worried about it but let's make sure everything is all right.
> 
> Regarding `bf16_full_eval`, it has been in my to-do list for a while but I have not got the time to deal with it yet. If you want to take a look at it, what needs to be done is the following:
> 
> * remove `bf16_full_eval` from the list [here](https://github.com/huggingface/optimum-habana/blob/94fe9152a4654b25af384dbf097f18ae57f22def/optimum/habana/transformers/training_args.py#L57) and remove the error [here](https://github.com/huggingface/optimum-habana/blob/94fe9152a4654b25af384dbf097f18ae57f22def/optimum/habana/transformers/training_args.py#L317)
> * take a look at the occurrences of `bf16_full_eval` in Transformers' Trainer (in particular at lines 1529, 3208 and 3816 [here](https://github.com/huggingface/transformers/blob/v4.34.1/src/transformers/trainer.py)) and replicate them in GaudiTrainer
> 
> That should be all that is needed but there might be a bit more, we should see this quite fast if any `bf16_full_eval` is missing anywhere else.

Thanks @regisss, I'll take a look at that once this is merged

**Comment by bhargaveede** [comment]
@regisss Can you merge this?

**Pull request merged**